src/v/cloud_storage/segment_meta_cstore.h

```diff
@@ -389,16 +389,9 @@ class segment_meta_column_const_iterator
     }
 
     bool equal(const auto& other) const {
-        // Invariant: _inner_it is never equal to _inner_end
-        // unless the iterator points to the end.
-        bool end_this = _inner_end == _inner_it;
-        bool end_other = other._inner_end == other._inner_it;
-        if (end_this && end_other) {
-            // All 'end' iterators are equal
-            return true;
-        }
-        return _outer_it == other._outer_it && _inner_it == other._inner_it;
-        return true;
+        ssize_t idx = is_end() ? -1 : index();
+        ssize_t oth_idx = other.is_end() ? -1 : other.index();
+        return idx == oth_idx;
```

Comment:
The test case test_segment_meta_cstore_iterators would fail before this commit. Only equality with end() would work, which is fine for iteration from a point to end, but it makes other patterns difficult to implement and does not match the ForwardIterator tag.